File: LinkedList/linkedList.cpp
#include <iostream>

using namespace std;

struct LinkedList
{
    int value;
    LinkedList *next;
};

LinkedList *head = NULL;
LinkedList *tail = head;

void insert(int value, int index = -1)
{
    LinkedList newNode;
    newNode.value = value;
    newNode.next = head;
    if (!head)
    {
        head = &newNode;
        tail = &newNode;
    }
    else
    {
        if (index != -1)
        {
            if (index == 0)
            {
                newNode.next = head;
                head = &newNode;
            }

            LinkedList *currentNode = head;
            LinkedList *parentNode = currentNode;
            int pos = 0;
            while (currentNode->next != NULL)
            {
                if (index == pos)
                {
                    newNode.next = parentNode->next;
                    parentNode->next = &newNode;
                }
                pos++;
            }
        }
        else
        {
            tail->next = &newNode;
        }
        tail = tail->next;
    }
}

void remove(int node)
{
    if (!head)
    {
        cout << "Linked List Empty" << endl;
        return;
    }
    if (node == head->value)
    {
        head = head->next;
        return;
    }

    LinkedList *currentNode = head;
    LinkedList *parentNode = currentNode;

    while (currentNode->next != NULL)
    {
        if (currentNode->value == node)
        {
            parentNode->next = currentNode->next;
            break;
        }
    }

    if (currentNode->value == tail->value)
    {
        tail = parentNode;
        tail->next = NULL;
    }
}

void print()
{
    LinkedList *currentNode = head;
    while (currentNode->next != NULL)
    {
        cout << currentNode->value << " -> ";
        currentNode = currentNode->next;
    }

    cout << currentNode->value << " -> " << endl;
}

int main()
{
    LinkedList ll;
    ll.value = 10;
    insert(25);
    insert(34);
    insert(12);
    insert(3);
    insert(87);
    insert(43);
    insert(29);
    insert(56);
    print();
    insert(89);
    remove(29);
    print();
    remove(10);
    remove(89);
    print();

    return 0;
}

File: LinkedList/linkedList.go
package main

import "fmt"

type LinkedList struct {
	value int
	next  *LinkedList
}

var head *LinkedList
var tail *LinkedList

func Init(value int) *LinkedList {
	l := &LinkedList{
		value: value,
		next:  nil,
	}
	head = l
	tail = l
	return l
}

func (l *LinkedList) Insert(value int) {
	newNode := LinkedList{
		value: value,
		next:  nil,
	}
	tail.next = &newNode
	tail = tail.next
}

func (l *LinkedList) Delete(node int) {
	currentNode := *head
	parentNode := currentNode
	if node == head.value {
		head = head.next
		return
	}
	for currentNode.next != nil {
		if currentNode.value == node {
			parentNode.next = currentNode.next
			break
		}
		parentNode = currentNode
		currentNode = *currentNode.next
	}
}

func (l *LinkedList) Print() {
	currentNode := *head
	for currentNode.next != nil {
		fmt.Printf("%d -> ", currentNode.value)
		currentNode = *currentNode.next
	}
	fmt.Println("")
}

func main() {
	ll := Init(10)
	ll.Insert(25)
	ll.Insert(34)
	ll.Insert(12)
	ll.Insert(3)
	ll.Insert(87)
	ll.Insert(43)
	ll.Insert(29)
	ll.Insert(56)
	ll.Print()
	ll.Insert(89)
	ll.Delete(29)
	ll.Print()
	ll.Delete(10)
	ll.Print()
}
